lib/endpoints/vnc.js

```diff
@@ -0,0 +1,304 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var jsprim = require('jsprim');
+var restify = require('restify');
+var schemas = require('joyent-schemas').cloudapi;
+var util = require('util');
+var vasync = require('vasync');
+var watershed = require('watershed');
+var mooremachine = require('mooremachine');
+var net = require('net');
+
+var shed = new watershed.Watershed();
+
+function mount(server, before, pre) {
+    assert.object(server, 'server');
+    assert.ok(before, 'before');
+    assert.optionalArrayOfFunc(pre, 'pre');
+
+    pre = pre || [];
+
+    server.get({
+        path: '/:account/machines/:machine/vnc',
+        name: 'ConnectMachineVNC',
+        version: [ '8.4.0' ]
+    }, before, connectVNC);
+
+    return server;
+}
+
+function connectVNC(req, res, next) {
+    var vm = req.machine;
+    if (vm.brand !== 'kvm') {
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'MachineHasNoVNC',
+            message: 'Specified machine does not have a VNC console'
+        }));
+        next();
+        return;
+    }
+    if (vm.state !== 'running') {
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'MachineStopped',
+            message: 'Specified machine is stopped and cannot be connected to'
+        }));
+        next();
+        return;
+    }
+    if (!res.claimUpgrade) {
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'UpgradeRequired',
+            message: 'VNC connect endpoint is a websocket and must be Upgraded'
+        }));
+        next();
+        return;
+    }
+
+    /*
+     * Since cloudapi still runs with restify request domains enabled, we need
+     * to exit that domain here if we want any errors in the VNC FSM to be
+     * reported sensibly (since the request will end from restify's
+     * perspective once we send the 101).
+     *
+     * This can be removed once domains and the uncaughtException handler are
+     * turned off for cloudapi.
+     */
+    var reqdom = process.domain;
+    reqdom.exit();
+
+    var fsm = new VNCConnectionFSM();
+    fsm.handle(req, res, next);
```

Comment:
This feels a little weird to me - I think I'd prefer to pass the arguments to the constructor, e.g.
```
    var fsm = new VNCConnectionFSM(req, res, next);
    fsm.handle();
```

Reply:
The only thing I can think of is that is might just be a common style w/ mooremachine that makes it easy to create the object at a time when those values aren't yet known, and then kick off the state machine at a later time once we're ready.  That doesn't appear to be necessary in this instance of course.

Reply:
That is not a typical convention for node (class instances), or mooremachine that I've seen (granted I have had limited expose to mooremachine):
https://github.com/joyent/sdc-net-agent/blob/master/lib/net-agent.js#L227
https://github.com/joyent/triton-cns/blob/d75aa32ebbba9cf8d655a17631aa9a5e00a341b2/lib/reaper-stream.js#L53
https://github.com/joyent/triton-cns/blob/74d5108c11ef76c0c8a9977065cd4c8100d1f34b/lib/cn-filter.js#L46

Reply:
I asked Alex, and he said he thought it might have been some cruft from a previous revision w/ more states. I've hopefully fixed that now.  `make check` wants something to use the `VNCConnectionFSM` instance after it's created, so I still left it so that you have to do 'something' (i.e. 'start') after you create the `VNCConnectionFSM` instance (happy to change it if you know of a better way).

Reply:
Using `fsm.start()` is fine - just for reference, I think the alternative (to keep make check happy) would have been to just drop the 'var fsm = ' variable assignment - since that fsm variable would not be used after that point - anyway, I'm fine with the separate .start() call.